superdesk/text_checkers/spellcheckers/languagetool.py

```diff
@@ -0,0 +1,166 @@
+# -*- coding: utf-8; -*-
+#
+# This file is part of Superdesk.
+#
+# Copyright 2013-2019 Sourcefabric z.u. and contributors.
+#
+# For the full copyright and license information, please see the
+# AUTHORS and LICENSE files distributed with this source code, or
+# at https://www.sourcefabric.org/superdesk/license
+
+import os
+import logging
+import requests
+import json
+from superdesk.errors import SuperdeskApiError
+from superdesk.text_checkers.spellcheckers import CAP_SPELLING, CAP_GRAMMAR, LANG_ANY
+from superdesk.text_checkers.spellcheckers.base import SpellcheckerBase
+
+logger = logging.getLogger(__name__)
+API_URL = "LANGUAGETOOL_API_URL"
+OPT_API_KEY = "LANGUAGETOOL_API_KEY"
+OPT_CONFIG = "LANGUAGETOOL_CONFIG"
+
+
+class Languagetool(SpellcheckerBase):
+    """Languagetool grammar/spellchecker integration
+
+    The LANGUAGETOOL_API_URL setting (or environment variable) must be set
+    """
+
+    name = "languagetool"
+    label = "Languagetool spellchecker"
+    capacities = (CAP_SPELLING, CAP_GRAMMAR)
+    languages = [LANG_ANY]
+
+    def __init__(self, app):
+        super().__init__(app)
+        self.api_url = self.config.get(API_URL, os.environ.get(API_URL))
+        if self.api_url:
+            self.api_url = self.api_url.rstrip('/')
+        self._languagetool_config = None
+
+    @property
+    def languagetool_config(self):
+        """Retrieve Languagetool config from settings.py or environment variables
+
+        config is cached once retrieved
+        """
+        if self._languagetool_config is None:
+            config = self.config.get(OPT_CONFIG, {})
+            api_key = self.config.get(OPT_API_KEY, os.environ.get(OPT_API_KEY))
+            if api_key:
+                config['apiKey'] = api_key
+            try:
+                env_config = json.loads(os.environ[OPT_CONFIG])
+            except (KeyError, json.JSONDecodeError):
+                env_config = {}
+            if not isinstance(config, dict) or not isinstance(env_config, dict):
+                logger.warning("Invalid type for {label} configuration, must be a dictionary".format(label=self.label))
+                self._languagetool_config = {}
+                return self._languagetool_config
+            config.update(env_config)
+            self._languagetool_config = config
+
+        return self._languagetool_config
+
+    def check(self, text, language=None):
+        payload = {
+            'text': text,
+            'language': language or 'auto',
+        }
+
+        # Add apiKey and additional options from the environment variable
+        additional_options = self.languagetool_config
+        payload.update(additional_options)
+
+        try:
+            # Send the POST request to the LanguageTool API
+            response = requests.post(self.api_url + '/check', data=payload)
```

Comment:
would be good to use a session, that will allow it to reuse connections

Reply:
Good point! Added now.